File: busca e ordenação/binary_search.c
// QUICK SORT

#include <stdio.h>

void quicksort(int lista[], int inicio, int fim){
    if (inicio < fim){
        int split;
        split = partition(lista, inicio, fim);
        quicksort(lista, inicio, split-1);
        quicksort(lista, split+1, fim);
    }
}


// BINARY SEARCH

int binary_search(int chave, int lista[], int inicio, int fim){
    int posicao = -1, meio;

    while (inicio <= fim){
        meio = (int)((inicio + fim) / 2);
        if (chave == lista[meio]){
            posicao = meio;
        } 
        if (chave < lista[meio]){
            fim = meio-1;
        }else{
            inicio = meio + 1;
        }
    }

    return posicao;
}


int main(){
    int lista[] = {89, 45, 68, 90, 29, 34, 17, 59}, chave, posicao;
    int n = sizeof(lista)/sizeof(lista[0]);

    quicksort(lista, 0, n-1);

    printf("\nDigite a chave: "); // Chave = K de Bob
    scanf("%d", &chave);

    posicao = binary_search(chave, lista, 0, n-1);
    
    printf("%d", lista[n-1-posicao]);

    return 0;
}

File: busca e ordenação/binary_search.py
'''BINARY SEARCH'''
from modules.utils import lista_aleatoria
from modules.insertion_sort import insertion_sort

def binary_search(chave, lista, inicio=0, fim=None):
    if (fim is None):
        fim = len(lista)-1

    posicao = -1
    while (inicio <= fim):
        meio = (inicio + fim)//2

        if (lista[meio] is chave):
            posicao = meio
        if (lista[meio] > chave):
            fim = meio-1
        else:
            inicio = meio+1

    return posicao


array = lista_aleatoria() # LEMBRAR DE ORDENAR O VETOR ANTES DE CHAMAR O BINARY SEARCH
array = insertion_sort(array)
print(array)
print([x for x in range(0, len(array))])
chave = int(input("Digite a chave: "))
posicao_chave = binary_search(chave, array)

if (posicao_chave != -1):
    print(f"A chave {chave} está na posição {posicao_chave}.")
else:
    print("A chave digitada não esá contida na lista.")
